mlir/include/mlir/Dialect/ArmSME/IR/ArmSMEOps.td

```diff
@@ -736,6 +736,7 @@ class OuterProductResultTileTypeConstraint<string operand> :
 
 def OuterProductOp :
   ArmSME_Op<"outerproduct", [
+    Pure,
```

Comment:
No ArmSME tile ops are pure, as they're implicitly working on ZA, which is not modeled. 

Consider: 
```
%0 = arm_sme.outerproduct %a, %b : vector<[4]x[4]xf32> // assume result is all 1 (allocs a new tile)
%1 = arm_sme.outerproduct %a, %b : vector<[4]x[4]xf32> // assume result is all 1 (allocs a new tile)
// Move tile slice into %1 (assume slice is all 0)
%2 = arm_sme.move_vector_to_tile_slice %1 ....
arm_sme.tile_store %0 ...
arm_sme.tile_store %2 ...
```
This results in a store of `%0` which is all 1, followed by a store of `%2`, which has one row of zeros.

Now since `arm_sme.outerproduct` is pure, cse can rewrite this to:
```
%0 = arm_sme.outerproduct %a, %b : vector<[4]x[4]xf32> // assume result is all 1
// Move tile slice into %1 (assume slice is all 0)
%1 = arm_sme.move_vector_to_tile_slice %0 ....
arm_sme.tile_store %0 ...
arm_sme.tile_store %1 ...
```
In this case all operations (after tile allocation) are working on the same tile. So _both_ will result in a store of a tile with one slice of zeros, which is incorrect.  

I don't think this can be fixed without proper tile value semantics.

Reply:
ah, good catch, I'm thinking of this op like a Vector dialect op and of course that's not the case. I'll remove this, it's not a problem because in the pass I'm explicitly checking all but the last outer product in a chain have single use and erasing them. Marking this as pure means cse would do that for us, but it's not necessary.